private/buf/cmd/buf/command/push/push.go

```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+//   - tags on current commit
+//   - current branch
+func getGitMetadataLabelsUploadOptions(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) (bufmodule.UploadOption, error) {
+	tags, err := getGitTagsOnCurrentCommit(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	branch, err := getCurrentGitBranch(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	labels := append(tags, branch)
+	if len(labels) > 0 {
+		return bufmodule.UploadWithLabels(labels...), nil
+	}
+	return nil, nil
+}
+
+func getGitTagsOnCurrentCommit(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) ([]string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("tag", "--points-at", "HEAD"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return nil, err
+	}
+	if len(buffer.Bytes()) > 0 {
+		return strings.Split(strings.TrimSpace(buffer.String()), "\n"), nil
+	}
+	return nil, nil
+}
+
+func getCurrentGitBranch(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) (string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("branch", "--show-current"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(buffer.String()), nil
+}
+
+func getGitMetadataSourceControlURLUploadOption(
+	ctx context.Context,
+	runner command.Runner,
+	remotes []string,
+	input string,
+) (bufmodule.UploadOption, error) {
+	remoteToURL := map[string]string{}
+	for _, remote := range remotes {
+		buffer := bytes.NewBuffer(nil)
+		if err := runner.Run(
+			ctx,
+			gitCommand,
+			command.RunWithArgs("config", "--get", fmt.Sprintf("remote.%s.url", remote)),
+			command.RunWithStdout(buffer),
+			command.RunWithStderr(buffer),
+			command.RunWithDir(input),
+		); err != nil {
+			return nil, err
+		}
+		remoteToURL[remote] = strings.TrimSpace(buffer.String())
+	}
+	// We prioritize the Git default remote, "origin", URL
+	if gitDefaultRemoteURL, ok := remoteToURL[gitDefaultRemote]; ok {
+		// First we need to trim the `.git` suffix if there is one.
+		gitDefaultRemoteURL = strings.TrimSuffix(gitDefaultRemoteURL, ".git")
+		// Then we get the current HEAD commit.
+		currentHEADCommit, err := getCurrentHEADGitCommit(ctx, runner, input)
+		if err != nil {
+			return nil, err
+		}
+		// Bitbucket is the only URL that uses the "/commits" route, both Github and Gitlab
+		// use "/commit", so we default to that if the remote URL does not point at Bitbucket.
+		if strings.HasPrefix(gitDefaultRemoteURL, bitBucketURL) {
+			return bufmodule.UploadWithSourceControlURL(fmt.Sprintf(
+				bitBucketRemoteURLFormat,
+				gitDefaultRemoteURL,
+				currentHEADCommit,
+			)), nil
+		}
+		return bufmodule.UploadWithSourceControlURL(fmt.Sprintf(
```

Comment:
Won't this break if someone is using a git remote instead of a https one? We should properly handle rewriting origins like `git@github.com:git/git.git` into appropriate URLs too.

Also I think we need to validate that the remote matches well known formats for GitHub/GitLab and if it doesn't match don't build a URL.